src/device_imxrt.c

```diff
@@ -0,0 +1,278 @@
+
+MG_IRAM void *mg_flash_start(void) {
+  return (void *) 0x0;
+}
+MG_IRAM size_t mg_flash_size(void) {
+  return 8 * 1024 * 1024;
+}
+MG_IRAM size_t mg_flash_sector_size(void) {
+  return 4 * 1024;  // 4k
+}
+MG_IRAM size_t mg_flash_write_align(void) {
+  return 256;
+}
+MG_IRAM int mg_flash_bank(void) {
+  return 0;
+}
+
+MG_IRAM static bool flash_page_start(volatile uint32_t *dst) {
+  char *base = (char *) mg_flash_start(), *end = base + mg_flash_size();
+  volatile char *p = (char *) dst;
+  return p >= base && p < end && ((p - base) % mg_flash_sector_size()) == 0;
+}
+
+// Note: the get_config function below works both for RT1020 and 1060
+#if MG_DEVICE == MG_DEVICE_RT1020
+MG_IRAM static int flexspi_nor_get_config(flexspi_nor_config_t *config) {
+  flexspi_nor_config_t default_config = {
+      .memConfig = {.tag = FLEXSPI_CFG_BLK_TAG,
+                    .version = FLEXSPI_CFG_BLK_VERSION,
+                    .readSampleClkSrc = 1,  // ReadSampleClk_LoopbackFromDqsPad
+                    .csHoldTime = 3,
+                    .csSetupTime = 3,
+                    .controllerMiscOption = MG_BIT(4),
+                    .deviceType = 1,  // serial NOR
+                    .sflashPadType = 4,
+                    .serialClkFreq = 7,  // 133MHz
+                    .sflashA1Size = 8 * 1024 * 1024,
+                    .lookupTable = __FLEXSPI_QSPI_LUT},
+      .pageSize = 256,
+      .sectorSize = 4 * 1024,
+      .ipcmdSerialClkFreq = 1,
+      .blockSize = 64 * 1024,
+      .isUniformBlockSize = false};
+
+  *config = default_config;
+  return 0;
+}
+#else
+MG_IRAM static int flexspi_nor_get_config(flexspi_nor_config_t *config) {
+  uint32_t options[] = {0xc0000000, 0x00};
+  if (!s_flash_irq_disabled) {
+    asm("cpsid i");
+  }
+  uint32_t status =
+      flexspi_nor->get_config(FLEXSPI_NOR_INSTANCE, config, options);
+  if (!s_flash_irq_disabled) {
+    asm("cpsie i");
+  }
+  if (status) {
+    MG_ERROR(("Failed to extract flash configuration: status %u", status));
+  }
+  return status;
+}
+#endif
+
+uint32_t mg_flash_init() {
+  flexspi_nor_config_t config;
+  if (flexspi_nor_get_config(&config) != 0) {
+    return false;
+  }
+  return flexspi_nor->init(FLEXSPI_NOR_INSTANCE, &config);
+}
+
+MG_IRAM bool mg_flash_erase(void *addr) {
+  flexspi_nor_config_t config;
+  if (flexspi_nor_get_config(&config) != 0) {
+    return false;
+  }
+  if (flash_page_start(addr) == false) {
+    MG_ERROR(("%p is not on a sector boundary", addr));
+    return false;
+  }
+  // Note: Interrupts must be disabled before any call to the ROM API on RT1020
+  if (!s_flash_irq_disabled) {
+    asm("cpsid i");
+  }
+  bool ok = (flexspi_nor->erase(FLEXSPI_NOR_INSTANCE, &config, (uint32_t) addr,
+                                mg_flash_sector_size()) == 0);
+  if (!s_flash_irq_disabled) {
+    asm("cpsie i");  // // Reenable them after the call
+  }
+  MG_DEBUG(("Sector starting at %p erasure: %s", addr, ok ? "ok" : "fail"));
```

Comment:
LOG from IRAM

Reply:
Normally, before the firmware swap, we make sure to disable the logs in order to make sure we don't call any log functions after erasing them. If logging is disabled, this macro has no effect.